File: sos_plugin/android/src/main/java/com/example/sos_plugin/SosPlugin.java
package com.example.sos_plugin;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class SosPlugin implements FlutterPlugin, MethodCallHandler {
    private MethodChannel channel;
    private InternetChecker internetChecker;
    private MotionChecker motionChecker;
    private LocationChecker locationChecker;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "sos_plugin");
        channel.setMethodCallHandler(this);

        internetChecker = new InternetChecker(flutterPluginBinding.getApplicationContext());
        motionChecker = new MotionChecker(flutterPluginBinding.getApplicationContext());
        locationChecker = new LocationChecker(flutterPluginBinding.getApplicationContext());

        // Setup Event Channels
        new EventChannel(flutterPluginBinding.getBinaryMessenger(), "sos_plugin/internet")
                .setStreamHandler(internetChecker);
        new EventChannel(flutterPluginBinding.getBinaryMessenger(), "sos_plugin/motionless")
                .setStreamHandler(motionChecker.getMotionlessStreamHandler());
        new EventChannel(flutterPluginBinding.getBinaryMessenger(), "sos_plugin/fall")
                .setStreamHandler(motionChecker.getFallStreamHandler());
        new EventChannel(flutterPluginBinding.getBinaryMessenger(), "sos_plugin/location")
                .setStreamHandler(locationChecker);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("setHomeLocation")) {
            double latitude = call.argument("latitude");
            double longitude = call.argument("longitude");
            locationChecker.setHomeLocation(latitude, longitude);
            result.success("Home location set");
        } else {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }
}

File: sos_plugin/ios/Classes/SosPlugin.swift
import Flutter
import UIKit
import CoreMotion
import CoreLocation

public class SosPlugin: NSObject, FlutterPlugin {
    private let internetChecker = InternetChecker()
    private let motionChecker = MotionChecker()
    private let locationChecker = LocationChecker()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "sos_plugin", binaryMessenger: registrar.messenger())
        let instance = SosPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Setup Event Channels
        let internetChannel = FlutterEventChannel(name: "sos_plugin/internet", binaryMessenger: registrar.messenger())
        internetChannel.setStreamHandler(instance.internetChecker)

        let motionlessChannel = FlutterEventChannel(name: "sos_plugin/motionless", binaryMessenger: registrar.messenger())
        motionlessChannel.setStreamHandler(instance.motionChecker.getMotionlessStreamHandler())

        let fallChannel = FlutterEventChannel(name: "sos_plugin/fall", binaryMessenger: registrar.messenger())
        fallChannel.setStreamHandler(instance.motionChecker.getFallStreamHandler())

        let locationChannel = FlutterEventChannel(name: "sos_plugin/location", binaryMessenger: registrar.messenger())
        locationChannel.setStreamHandler(instance.locationChecker)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setHomeLocation":
            if let args = call.arguments as? [String: Double],
               let latitude = args["latitude"],
               let longitude = args["longitude"] {
                locationChecker.setHomeLocation(latitude: latitude, longitude: longitude)
                result("Home location set")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Latitude or Longitude missing", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}